### ISSUE TITLE
[0.16] Simple fix for missing Decorator support

### ISSUE THREAD

**Issue by idibidiart** [facebook/react-native#4577]
I had to overwrite react-packager's babelrc file to add transform-decorators-legacy plugin before the classes plugin to get decorators working again. At first, I supplied my own babelrc to override the one that's in the react-packager folder but that didn't work, so I no longer use a babelrc at the project level, which means only way to use legacy decorators with RN 0.16.0 is to fork RN by making this modification 

```
{
  "retainLines": true,
  "compact": true,
  "comments": false,
  "plugins": [
    "syntax-async-functions",
    "syntax-class-properties",
    "syntax-trailing-function-commas",
    "transform-class-properties",
    "transform-es2015-arrow-functions",
    "transform-es2015-block-scoping",
    "transform-decorators-legacy",
    "transform-es2015-classes",
    "transform-es2015-computed-properties",
    "transform-es2015-constants",
    "transform-es2015-destructuring",
    ["transform-es2015-modules-commonjs", {"strict": false, "allowTopLevelThis": true}],
    "transform-es2015-parameters",
    "transform-es2015-shorthand-properties",
    "transform-es2015-spread",
    "transform-es2015-template-literals",
    "transform-flow-strip-types",
    "transform-object-assign",
    "transform-object-rest-spread",
    "transform-react-display-name",
    "transform-react-jsx",
    "transform-regenerator",
    "transform-es2015-for-of"
  ],
  "sourceMaps": false
}
```

**Comment by idibidiart** [comment]
Seems like I didn't even have to do this.

Simply having these as dependencies in my package.json (in root of app) seems to have added support for decorators
```
  "dependencies": {
    "babel-plugin-transform-runtime": "*",
    "babel-preset-stage-0": "*",
    "babel-preset-stage-1": "*",
    "babel-preset-stage-2": "*",
    "babel-preset-stage-3": "*",
    "babel-preset-es2015": "*",
    "babel-preset-react": "*",
    "react-native": "0.16.0",
...
```

**Comment by idibidiart** [comment]
Nope, my NPM cache was dirty.

So in the end, the thing that worked was to modify React Packager's babelrc to add the decorators-legacy transform

```{
  "retainLines": true,
  "compact": true,
  "comments": false,
  "plugins": [
    "syntax-async-functions",
    "syntax-class-properties",
    "syntax-trailing-function-commas",
    "transform-class-properties",
    "transform-es2015-arrow-functions",
    "transform-es2015-block-scoping",
    "transform-decorators-legacy",
    "transform-es2015-classes",
    "transform-es2015-computed-properties",
    "transform-es2015-constants",
    "transform-es2015-destructuring",
    ["transform-es2015-modules-commonjs", {"strict": false, "allowTopLevelThis": true}],
    "transform-es2015-parameters",
    "transform-es2015-shorthand-properties",
    "transform-es2015-spread",
    "transform-es2015-template-literals",
    "transform-flow-strip-types",
    "transform-object-assign",
    "transform-object-rest-spread",
    "transform-react-display-name",
    "transform-react-jsx",
    "transform-regenerator",
    "transform-es2015-for-of"
  ],
  "sourceMaps": false
}```

This works after clearing NPM cache etc

**Comment by sjmueller** [comment]
@ide, can you give some advice on how to best configure a react-native 0.16 project with legacy decorator support?

**Comment by ide** [comment]
@sjmueller I don't have a great solution -- waiting on some upstream fixes for the babel plugin, and then it's going to require a fork of React Native.

**Comment by sjmueller** [comment]
Roger that. I was under the impression that there was a good way, and we were just doing it wrong.  I'll wait awhile before upgrading from 0.15, because I can't live without decorators anymore :astonished:

**Comment by idibidiart** [comment]
@sjmueller @ide i was able to get decorators working in 0.16 by adding transform-decorators-legacy to react-packager's babelrc before the classes transform and by also adding babel-plugin-transform-decorators-legacy to my app's package.json

I consider this a simple solution but it does mean that i have to overwrite react-packager's babelrc after i do npm install. I also no longer have my own babelrc.

**Comment by skevy** [comment]
@idibidiart I am also familiar with the issue that the projects .babelrc doesn't seem to be respected correctly. Rather than continuing on on this issue, could you open another issue about that problem specifically?

Re: decorators...babel-transform-decorators-legacy is a good solution for now, plus, it's logged in breaking-changes.md (https://github.com/facebook/react-native/blob/master/breaking-changes.md)

**Comment by idibidiart** [comment]
@skevy will do! thx

**Comment by cancan101** [comment]
@idibidiart Would you link to the new issue?

**Comment by sjmueller** [comment]
So is it possible yet to use `babel-transform-decorators-legacy` without modifying react-packager's babelrc inside of `node_modules`?  If so, please illustrate.

**Comment by sjmueller** [comment]
For the record, I've tried to simply add .babelrc `{ plugins: ["transform-decorators-legacy"] }` and `npm i --save-dev babel-plugin-transform-decorators-legacy` to my RN 0.17 project.  Compilation then seems to understand the syntax (not throwing an 'unexpected token' exception), however my class level decorators are not executed at runtime.

**Comment by hswolff** [comment]
I got decorators to work with 0.17.0 yesterday through slightly unorthodox means, however it's a stop-gap until it's supported upstream.

1. Install the decorator plugin locally via `npm install --save-dev babel-plugin-transform-decorators-legacy`

2. In your `package.json` file add a `postinstall` command to run a small script I made to patch react-native's `.babelrc` file to include support for the legacy decorator support.

```
"postinstall": "node scripts/patch-react-native-babelrc.js",
```

Contents of `patch-react-native-babelrc.js`:

```js
/**
 * This file patches react-native's packager's .babelrc file to add support
 * for decorators. There's no way to subclass and extend the .babelrc file,
 * as such we have to directly modify it.
 */

const path = require('path');
const fs = require('fs');

// Path to react-native's .babelrc file.
// This value can change as we update react-native.
const babelRcPath = path.resolve(
  __dirname,
  '../node_modules/react-native/',
  'packager/react-packager/.babelrc'
);


const babelRc = fs.readFileSync(babelRcPath, 'utf8');
const babelRcJson = JSON.parse(babelRc);

// Decorator plugin we want to add to react-native's packager.
const decoratorPlugin = 'transform-decorators-legacy';

// If we haven't already modified the .babelrc file then do it now.
if (babelRcJson.plugins[0] !== decoratorPlugin) {
  console.log('Patched react-native packager .babelrc');

  // Prepend.
  babelRcJson.plugins.unshift(decoratorPlugin);

  // Write file.
  fs.writeFileSync(
    babelRcPath,
    JSON.stringify(babelRcJson),
    'utf8'
  );
}
```

Completely unorthodox but it works.

**Comment by idibidiart** [comment]
Well the way I have been doing it is simply by ignoring node modules whole excluding React Packager's. Babelrc in my .gitignore so this way I simply checkout it out from the repo after nom install. No need to run any script. 

That's been working fine as long as i remember to do that after npm install. 

If anyome can start another issue about user's babelrc needing to be applied to user project rather than react's own babelrc that woild be beneficial i think. If uou do, please reference this issue.

**Comment by idibidiart** [comment]
*after npm install*

**Comment by idibidiart** [comment]
I no longer need to use the autobind decorator, and that was my main use case. I'm using property initializers with the arrow function syntax to get around having to explicitly bind to component instance.

here is the babel example:

http://bit.ly/1UIeMWw

**Comment by TossShinHwa** [comment]
@hswolff , interesting solution, helpful.

But we hope we can custom ourself's `.babel` file, not which in `node_modules`'s one. I guess `react-native` shouldn't include `babel` and `.babelrc`.

**Comment by hswolff** [comment]
So as I currently understand it, `react-native`'s packager will include your local `.babelrc` file and append it to their own `.babelrc` file. That's great, but in the case of decorators, it requires that it is placed before other transform plug-ins, rather than being appended to the end. The solution to allowing for dictating order is a little tricky, and letting people use only their own `.babelrc` could lead to other issues where `react-native` source code is transformed incorrectly.

Easiest solution IMHO is to just add decorator support upstream.

**Comment by tomprogers** [comment]
Thanks, @hswolff , for that fantastic postinstall hook. 8)

**Comment by hswolff** [comment]
Right you are @tomprogers, already updated it to work with 0.20.0.

Only pre-requisite is to either use npm v3 OR `npm install --save-dev babel-preset-react-native` so that the path to the file is consistent.

Here's a working version for 0.20.0:

```js
/**
 * This file patches react-native's packager's preset to add support
 * for decorators. There's no way to control the order in which plugins
 * are applied so we have to directly modify the file.
 */

const path = require('path');
const fs = require('fs');

// Path to config file.
// This value can change as we update react-native.
const mainConfigPath = path.resolve(
  __dirname,
  '../node_modules/babel-preset-react-native/',
  'configs/main.js'
);

var mainConfigContent = fs.readFileSync(mainConfigPath, 'utf8');

const prependToTransform = '\'transform-class-properties\',';

// Decorator plugin we want to add to react-native's packager.
const decoratorPlugin = '\'transform-decorators-legacy\',';

// If we haven't already modified the .babelrc file then do it now.
if (mainConfigContent.indexOf(decoratorPlugin) === -1) {
  console.log('Patched babel-preset-react-native main.js');

  // Prepend.
  mainConfigContent = mainConfigContent.replace(
    prependToTransform,
    decoratorPlugin + prependToTransform
  );

  // Write file.
  fs.writeFileSync(
    mainConfigPath,
    mainConfigContent,
    'utf8'
  );
}
```

It's even more hacky but it still works :).

**Comment by skevy** [comment]
@everyone 

In 0.20, all you should have to do is the following:

```
npm install --save babel-preset-react-native babel-plugin-transform-decorators-legacy
```

And then in a .babelrc in your project root, you should be able to do:

```
{
  "presets": ["react-native"],
  "plugins": ["transform-decorators-legacy"]
}
```

This should work on NPM 2 and NPM 3. No hacks should be required as they were before -- simplifying this is exactly why we built the preset.

If it does not, then please get back to me with reproducible steps and we will get this fixed. In those reproducible steps, please provide a gist with the output of `npm ls`, the contents of your babelrc, and the contents of your package.json.

Thanks.

**Comment by hswolff** [comment]
@skevy I tried that originally, however the issue is that `babel-plugin-transform-es2015-classes` throws an error if it encounters a decorator, asking you to place the decorator plugin before the class plugin. As far as I know you can't dictate order of plugins with a .babelrc file.

The error that gets thrown is found here: https://github.com/babel/babel/blob/93e5c0e64b1a14f3b138a01c55082225084f47b4/packages/babel-plugin-transform-es2015-classes/src/vanilla.js#L230

**Comment by skevy** [comment]
You're correct that you can't specify order...it's interesting though because this works for me on my projects where I'm using class decorators.

Do you have a small test case that breaks? Or could you make one? (Maybe on a new `react-native init`'ed project?)

**Comment by jeanregisser** [comment]
I confirm @skevy's recommandation works here as well.

**Comment by jsierles** [comment]
Looks like this issue is solvable using the various suggestions in this thread, so I'm closing this out to keep our issue count down.

Feel free to discuss more here or reopen if this issue persists!

**Comment by shahankit** [comment]
@jsierles @skevy I still cannot get decorators working on 0.22.0 I created fresh project using `react-native init testApp` and following is my dependencies in package.json:

```
"autobind-decorator": "^1.3.3",
"babel-core": "^6.7.4",
"babel-loader": "^6.2.4",
"babel-plugin-transform-class-properties": "^6.6.0",
"babel-plugin-transform-decorators-legacy": "^1.3.4",
"babel-plugin-transform-runtime": "^6.6.0",
"babel-preset-es2015": "^6.6.0",
"babel-preset-react": "^6.5.0",
"babel-preset-react-native": "^1.5.6",
"babel-preset-stage-0": "^6.5.0",
"babel-runtime": "^6.6.1",
"react": "^0.14.8",
"react-native": "^0.22.2",
"webpack": "^1.12.14"
```

and following is my babelrc:
```
{
  "plugins": [
    "transform-runtime",
    "transform-decorators-legacy",
    "transform-class-properties"
  ],
  "presets": ["es2015", "react", "stage-0"]
}
```

**Comment by mmazzarolo** [comment]
@shahankit are all those presets needed?  
If I remember correctly the presets order declaration is important, maybe try removing some presets and see if you can isolate the issue.  
Decorators works fine for me with the following package/babelrc:

```json
"babel-eslint": "^6.0.0",
"babel-plugin-transform-decorators-legacy": "^1.3.4",
"babel-preset-react-native": "^1.5.6",
```

```json
{
  "presets": ["react-native"],
  "plugins": ["transform-decorators-legacy"]
}
```

**Comment by shahankit** [comment]
@mmazzarolo Can you please tell me what else do you have in you package.json regarding babel. I reduced my .babelrc to exactly above 2 lines but now giving me error at line of decorator. I've pushed a demo app [here](https://github.com/shahankit/testApp)

**Comment by mmazzarolo** [comment]
> @mmazzarolo Can you please tell me what else do you have in you package.json regarding babel. I reduced my .babelrc to exactly above 2 lines but now giving me error at line of decorator. I've pushed a demo app here  

Of course:
```json
{
  "name": "Sensi",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "reset": "watchman watch-del-all && rm -rf node_modules/ && npm cache clean && npm prune && npm i",
    "generate-apk": "curl 'http://localhost:8081/index.android.bundle?platform=android' -dev=false -o 'android/app/src/main/assets/index.android.bundle' && android/gradlew assembleRelease -p android/",
    "lint": "standard --verbose | snazzy"
  },
  "dependencies": {
    "autobind-decorator": "^1.3.3",
    "react": "^0.14.8",
    "react-native": "^0.22.2",
    "react-native-animatable": "^0.5.2",
    "react-native-beacons-android": "^1.0.0",
    "react-native-i18n": "0.0.8",
    "react-native-router-flux": "^3.2.6",
    "react-redux": "^4.4.1",
    "redux": "^3.3.1",
    "redux-logger": "^2.6.1",
    "seamless-immutable": "^5.1.1"
  },
  "devDependencies": {
    "babel-eslint": "^6.0.0",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-preset-react-native": "^1.5.6",
    "eslint-plugin-flowtype": "^2.2.2",
    "snazzy": "^3.0.0",
    "standard": "^6.0.8"
  },
  "standard": {
    "parser": "babel-eslint",
    "plugins": [
      "flowtype"
    ],
    "globals": [
      "__DEV__",
      "fetch",
      "Parse"
    ]
  }
}
```

But you can also find a working setup [here](https://github.com/mmazzarolo/google-play-music-desktop-remote/tree/c8063fe58ff748dcc79db13742dcc38ab91323c2)

**Comment by badasve** [comment]
After some time trying I found out that @skevy 's solution only works if you do put the plugin under the "plugins" key of a .babelrc file at the root of your projet, not if you put it in your Package.json under the "babel" key

**Comment by badasve** [comment]
After some time trying I found out that @skevy 's solution only works if you do put the plugin under the "plugins" key of a .babelrc file at the root of your projet, not if you put it in your Package.json under the "babel" key, for me the react-native preset was not needed, using RN0.34
